DOCS-3430: Add clarification on scenario for resource level policies … (#297)
mulesoft/docs-api-manager: modules/ROOT/pages/gatekeeper.adoc

endif::[]
:keywords: runtime, gateway, gatekeeper, security, policy, policies

> When you deploy an application, the tracked resources exposed by the application might become accessible for a short amount of time due to a lapse in security. Because the policies specified for those tracked resources are not yet applied, the API might receive unwanted traffic. GateKeeper prevents this situation by blocking the tracked resource until all policies are retrieved and applied without errors. 

Reviewer: ```suggestion
When you deploy an application, the tracked resources exposed might become accessible for a short amount of time. Because the policies for those resources are not yet applied, the API might receive unwanted traffic. GateKeeper prevents this situation by blocking the tracked resource until all policies are retrieved and applied without errors. 
```
"policies" = "security policies"

Reviewer 2: Leaving it as is. The previous text explained why the lapse occurs. That is needed. Done! Thanks!

== How Gatekeeper Works

> The Gatekeeper mechanism only applies when you (re)start the runtime or (re)deploy a tracked resource that is linked with an API instance through Autodiscovery. When Gatekeeper is engaged, each tracked resource is blocked ("503 - Service Unavailable" HTTP status code) until all API instance policies  are applied in API Manager without errors.

Reviewer: ```suggestion
The Gatekeeper mechanism is engaged only when you restart the runtime or redeploy a tracked resource that is linked with an API instance through Autodiscovery. When Gatekeeper is engaged, each tracked resource is blocked, resulting in a `503 - Service Unavailable` HTTP status code, until all API instance policies are applied in API Manager without errors.
```
Just FYI, I looked up Autodiscovery and found several different uses in our legacy doc (API Autodiscovery, Auto-Discovery, the Auto-Discovery module, the Autodiscover service, autodiscovery...). I have a question out to the PMMs about the proper naming.

Reviewer 2: Done! Thanks!

> Resources without an associated Autodiscovery are ignored by Gatekeeper. Additionally, you cannot apply any policies to these resources and no analytics information is generated. If you loose connectivity after the Mule instance is (re)started or the tracked Mule application is (re)deployed, you use cached policies, configuration, and contracts even if Gatekeeper is not enabled. 

Reviewer: ```suggestion
Resources not linked to an API instance through Autodiscovery are ignored by Gatekeeper. Additionally, you cannot apply any policies to these resources, and no analytics information is generated. If you lose connectivity after the Mule instance is restarted or the tracked Mule application is redeployed, you use cached policies, configuration, and contracts, even if Gatekeeper is not enabled. 
```
Some of the rewording edits are an attempt to use just one wording rather than different wordings for the same thing, for example, Gatekeeper being "engaged" and "resources linked to an API instance through Autodiscovery".

As a reader, I don't know what you are implying by using "(re)"? I suggest that you choose the "re" version if that is most likely, or spell out the whole thing so that meaning is crystal clear: started or restarted, deployed or redeployed....

Reviewer 2: Done! Thanks!

> If a malformed or incorrectly defined policy is to be applied to a tracked resource and Gatekeeper is engaged in this scenaroio, the tracked resource is blocked with "503 - Service Unavailable" HTTP status code until the offending policy is removed from API Manager. The Mule runtime engine locally caches all policies and active contracts.

Reviewer: Building on the previous paragraph...:
```suggestion
If Gatekeeper is engaged and a tracked resource has a malformed or incorrectly defined policy applied, the tracked resource is blocked, indicated by the `503 - Service Unavailable` HTTP status code, until the offending policy is removed from API Manager. The Mule runtime engine locally caches all policies and active contracts.
```

Reviewer 2: Done! Thanks!

endif::[]
:keywords: runtime, gateway, gatekeeper, security, policy, policies

> When you deploy an application, the tracked resources exposed by the application might become accessible due to a lapse. Because the policies specified for those tracked resources are not yet applied, the API might receive unwanted traffic. GateKeeper prevents this situation by blocking the tracked resource until all policies are retrieved and applied without errors. 

Author: due to a lapse. Seems that the sentence is incomplete

Reviewer 2: Done! Thanks!

== How Gatekeeper Works

> The Gatekeeper mechanism only applies when you (re)starting the runtime or (re)deploying a tracked resource that is linked with an API instance through Autodiscovery. When Gatekeeper is engaged, each tracked resource is blocked ("503 - Service Unavailable" HTTP status code) until all policies that are applied to the API instance in API Manager are run without errors.

Author: you starting seems wrong

Author: API Manager are run without errors ==> API Manager are applied without errors

Reviewer 2: Done! Thanks!

Resources without an associated Autodiscovery are ignored by Gatekeeper. Additionally, you cannot apply any policies to these resources and no analytics information is generated.

> If you loose connectivity after the Mule instance is (re)started or the tracked Mule application is (re)deployed, you use cached policies, configuration and contracts with a different mechanism, during the connectivity outage. This mechanism is implemented in all versions of the Mule instances that support API gateway capabilities.

Author: I don't like the semantics of this part. What we should mention is that API Gateway will still use cached policies, configuration and contracts even if Gatekeeper is not enabled. The different mechanism clarification does not give any value  think.

Reviewer 2: Done! Thanks!

=== Security Level Definitions

> You can apply one of the listed security levels to Gatekeepr:

Reviewer: Typo (Gatekeepr) and also a colon should not precede a subhead.

FIRST SUGGESTION...but I now think that this should be a table, or a list that pairs the equivalent levels: `false` and `disabled`, `true` and `strict`, and `flexible`:
_I suggest you intro the security levels in this opening sentence first, and then precede to the subtopics for each.
Alternatively, you could make this a definition list or a bullet list instead of using subtopics._

Reviewer: See below

Reviewer 2: Done! Thanks!

==== False

> When you set this value, Gatekeeper does not apply any security. With this setting, the API might receive unwanted traffic during the period while the policies are being applied. In Mule 3.x, this value is available only for backward compatibility. Therefore, use ‘disabled’ instead for all other scenarios. Note that the `disabled` value is not available in all Mule 3.x releases and the `false` value is not available in Mule 4.x releases.

Reviewer: ```suggestion
When you set this value, Gatekeeper does not apply any security, and the API might receive unwanted traffic before the policies are completely applied, without errors. In Mule 3.x, `false` is available only for backward compatibility and is replaced by `disabled` in Mule 4.x.
```

Reviewer 2: Done! Thanks!

==== Disabled

> When you set this value, Gatekeeper does not apply any security. With this setting, the API might receive unwanted traffic during the period while the policies are being applied.

Reviewer: This should use exactly the same first sentence as the previous value's definition (minus the versioning info).

Reviewer: ...but now I think we should pair the two rather than separating them....
See below.

Reviewer 2: Done! Thanks!

==== True

When you set this value, the default security level is  set to `Strict`. In Mule 3.x, this value is available only for backward compatibility. Therefore, use ‘strict’ or ‘flexible’ instead for all other scenarios. The value `true` is unavailable in Mule 4.x releases.

> 

Reviewer: I think all this narrative is complicating the point that security levels used in Mule 3.x are replaced in Mule 4.x.
It might be more clear to use a table that defines the actual security and then has a column each for value in Mule 3 and Mule 4, something like...

**Type of Security-----Definition-------------Level in Mule 3.x-----       Level in Mule 4.x-----**

None-------------------blah blah-----------------`false`------- -----------`disabled`

Reviewer 2: Done! Thanks!

==== Strict

> When you set this value, Gatekeeper security is set on its maximum security level.

Reviewer: ```suggestion
When you set this value, Gatekeeper security is set to its maximum security level.
```

Reviewer 2: Done! Thanks!

> When the Mule instance is (re)started, the API is blocked and a "503 - Service Unavailable" HTTP status code is returned until policies for each tracked resource are correctly applied. If the connectivity to Anypoint Platform is lost when Gatekeeper is engaged, the tracked resource is blocked until the connection is reestablished.

Reviewer: ```suggestion
When the Mule instance is restarted, the API is blocked with `503 - Service Unavailable` until policies for each tracked resource are correctly applied. If connectivity to Anypoint Platform is lost when Gatekeeper is engaged, the tracked resource is blocked until the connection is reestablished.
```
I think we can shorten the text around 503 now that it has been established....

Reviewer 2: Done! Thanks!

==== Flexible

When this value is established, Gatekeeper security is set on its intermediate security level.

> When the Mule instance is (re)started, the API is blocked and a "503 - Service Unavailable" HTTP status code is returned until policies for each tracked resource are correctly applied. If the connectivity to Anypoint Platform is lost when Gatekeeper is engaged, the local cache is used to apply the policies during the outage. The tracked resource is blocked at least until the connection is reestablished if cache is not available.

Reviewer: ```suggestion
When the Mule instance is restarted, the API is blocked and `503 - Service Unavailable` until policies for each tracked resource are correctly applied. If the connectivity to Anypoint Platform is lost when Gatekeeper is engaged, local cache is used to apply the policies during the outage. If cache is unavailable, The tracked resource is blocked at least until the connection is reestablished.
```
"at least"? what would cause it to continue to be blocked?

Reviewer 2: Done! Thanks!

The `Flexible` mode is available in Mule versions v3.8.6 or later, v3.8.x, v3.9.0 or later ,v3.9.x, and v4.1.0 or later.

> === Default Gatekeeper Configuration

Reviewer: So I am thinking the whole section could be replaced by an expanded version of this table, with all the versions, default levels, and definitions of that those levels mean.

Reviewer 2: Added a table with column for 3.x and 4.x.

=== Security Level Definitions

You can apply one of the listed security levels to Gatekeepr:

==== False

> When you set this value, Gatekeeper does not apply any security. For that reason, the API might receive unwanted traffic during the period when the policies are applied. In Mule 3.x, this value is available only for backward compatibility. Therefore, Use ‘disabled’ instead for all other scenarios. Note that the `disabled` value is not available in all Mule 3.x releases and the value false` is not available in Mule 4.x releases.

Author: while policies are being applied

Reviewer 2: Done! Thanks!

==== Disabled

> When you set this value, Gatekeeper does not apply any security. For that reason, the API might receive unwanted traffic during the period when the policies are applied.

Author: Same

Reviewer 2: Done! Thanks!

==== True

When you set this value, the default security level is  set to `Strict`. In Mule 3.x, this value is available only for backward compatibility. Therefore, use ‘strict’ or ‘flexible’ instead for all other scenarios. The value `true` is unavailable in Mule 4.x releases.

==== Strict

When you set this value, Gatekeeper security is set on its maximum security level.

When the Mule instance is (re)started, the API is blocked and a "503 - Service Unavailable" HTTP status code is returned until policies for each tracked resource are correctly applied. If the connectivity to Anypoint Platform is lost when Gatekeeper is engaged, the tracked resource is blocked until the connection is reestablished.

==== Flexible

When this value is established, Gatekeeper security is set on its intermediate security level.

> When the Mule instance is (re)started, the API is blocked and a "503 - Service Unavailable" HTTP status code is returned until policies for each tracked resource are correctly applied. If the connectivity to Anypoint Platform is lost when Gatekeeper is engaged, a local cache is used to download and apply the policies during the outage. The tracked resource is blocked at least until the connection is reestablished.

Author: Last sentence should clarify that is blocked until connection is reestablished only if cache is not present.
Also, there is no download when cache is used, so I would just say local cache is used to apply the policies during the outage

Reviewer 2: Done! Thanks!

s| 4.1.0  and later | Flexible
|===

== Troubleshooting Gatekeeper Issues in Mule v3.x

==== Gatekeeper does not unblock the API when a resource level policy is applied to it 

> In Mule 3, you can only apply resource-level policies to a RAML proxy or to an API implementation containing an APIKit router. Gatekeeper might repeatedly block an API when a resource-level policy is applied to it due to two main reasons:

Reviewer: ```suggestion
In Mule 3.x, you can apply resource-level policies only to a RAML proxy or to an API implementation containing an APIKit router. Gatekeeper might repeatedly block an API when a resource-level policy is applied to it for one of the following reasons:
```

Reviewer 2: Done! Thanks!

> * The selected expressions for URI and method when configuring the resource-level policy do not match with any resource specified in the RAML definition. To avoid this type of error, use the `Preview Resource Matching` option in API Manager when configuring the specific resource-level policy. 

Reviewer: Please use proper bullet-list format:
```suggestion
* The expressions for URI and method selected when configuring the resource-level policy do not match any resource specified in the RAML definition. 
+
To avoid this type of error, use the `Preview Resource Matching` option in API Manager when configuring the specific resource-level policy. 
```

Reviewer 2: Done! Thanks!

> * The autodiscovery element defined in the Mule application does not specify the `apikitRef` attribute. To fix this issue, add the name of the APIKit configuration that the router is using to the 'apikitRef' attribute.

Reviewer: ```suggestion
* The autodiscovery element defined in the Mule application does not specify the `apikitRef` attribute. 
+
To fix this issue, add the name of the APIKit configuration that the router is using to the 'apikitRef' attribute.
```

Reviewer 2: Done! Thanks!
